src/coord/src/coord.rs

```diff
@@ -1664,7 +1664,8 @@ where
         let view_id = self.catalog.allocate_id()?;
         let view_oid = self.catalog.allocate_oid()?;
         // Optimize the expression so that we can form an accurately typed description.
-        let optimized_expr = self.optimizer.optimize(view.expr, self.catalog.indexes())?;
+        let ts = None; // views don't execute at a logical timestamp
+        let optimized_expr = self.optimize_expr(view.expr, ts)?;
```

Comment:
If optimizing an expression without a logical timestamp substitution will fail, why doesn't this fail too? The `optimize_expr` code seems to optimize the expression even without substituting a time when `None` is supplied.

Reply:
It's because the SQL planner rejects `mz_logical_timestamp` calls that occur in a `CREATE VIEW` statement. So it was ok not to replace here, because there couldn't possibly be any exprs to replace. But not in the other code path.

I decided to rework this more to make it less confusing. Now the SQL planner is always willing to plan mz_logical_timestamp, and the coordinator is in charge of deciding whether it can perform a replacement or not.